dnstable/entry.c

```diff
@@ -82,56 +82,37 @@ fmt_uint64_str(char *s, uint64_t u)
 static void
 fmt_time(ubuf *u, uint64_t v)
 {
-	struct tm gm;
+	struct tm gm, *r;
 	time_t tm = v;
 	char s[sizeof("4294967295-12-31 23:59:59 -0000")];
-	if (gmtime_r(&tm, &gm) != NULL) {
-		snprintf(s, sizeof(s), "%d-%02d-%02d %02d:%02d:%02d -0000",
-			1900 + gm.tm_year,
-			1 + gm.tm_mon,
-			gm.tm_mday,
-			gm.tm_hour,
-			gm.tm_min,
-			gm.tm_sec
-	       );
-	}
-	ubuf_add_cstr(u, s);
+	r = gmtime_r(&tm, &gm);
+	if ((r != NULL) && (strftime(s, sizeof(s), "%Y-%m-%d %H:%M:%S -0000", r) > 0))
+		ubuf_add_cstr(u, s);
 }
 
 static void
 fmt_rfc3339_time(ubuf *u, uint64_t v)
 {
-	struct tm gm;
+	struct tm gm, *r;
 	time_t tm = v;
 	char s[sizeof("4294967295-12-31T23:59:59Z")];
 
-	assert (gmtime_r(&tm, &gm) != NULL);
-	snprintf(s, sizeof(s), "%d-%02d-%02dT%02d:%02d:%02dZ",
-		 1900 + gm.tm_year,
-		 1 + gm.tm_mon,
-		 gm.tm_mday,
-		 gm.tm_hour,
-		 gm.tm_min,
-		 gm.tm_sec
-		);
-	ubuf_add_cstr(u, s);
+	r = gmtime_r(&tm, &gm);
+	if ((r != NULL) && (strftime(s, sizeof(s), "%Y-%m-%dT%H:%M:%SZ", r) > 0))
+		ubuf_add_cstr(u, s);
 }
 
 static int
 fmt_rfc3339_time_str(uint64_t v, char *ts, size_t len_ts)
 {
-	struct tm gm;
+	struct tm gm, *r;
 	time_t tm = v;
 
-	assert (gmtime_r(&tm, &gm) != NULL);
-	return snprintf(ts, len_ts, "%d-%02d-%02dT%02d:%02d:%02dZ",
-			1900 + gm.tm_year,
-			1 + gm.tm_mon,
-			gm.tm_mday,
-			gm.tm_hour,
-			gm.tm_min,
-			gm.tm_sec
-		);
+	r = gmtime_r(&tm, &gm);
+	if (r == NULL)
+		return 0;
+
+	return strftime(ts, len_ts, "%Y-%m-%dT%H:%M:%SZ", r);
```

Comment:
The return value from strftime in case of errors is different than from snprintf.
Using the macOS docs:
- snprintf()  returns the number of characters that would have been printed if the size were unlimited (again, not including the final `\0').  Return a negative value if an error occurs.
-  strftime() returns the number of characters in the array, not counting the terminating NUL.  Otherwise, zero is returned.

But the two uses of the private fmt_rfc3339_time_str() function at lines 310 and 328 don't check the return value anyway.  Thus I will accept the changes, but we "should" fix 310 and 328.